internal/pkg/diagnostics/diagnostics.go

```diff
@@ -89,14 +102,166 @@ func GlobalHooks() Hooks {
 		if err != nil {
 			// error is returned as the content
-			return []byte(fmt.Sprintf("failed to write pprof to bytes buffer: %s", err))
+			return []byte(fmt.Sprintf("failed to write pprof to bytes buffer: %s", err)), time.Now().UTC()
+		}
+		return w.Bytes(), time.Now().UTC()
+	}
+}
+
+// ZipArchive creates a zipped diagnostics bundle using the passed writer with the passed diagnostics.
+// If any error is encountered when writing the contents of the archive it is returned.
+func ZipArchive(errOut, w io.Writer, agentDiag []client.DiagnosticFileResult, unitDiags []client.DiagnosticUnitResult) error {
+	zw := zip.NewWriter(w)
+	defer zw.Close()
+	// Create directories in the zip archive before writing any files
+	for _, ad := range agentDiag {
+		if ad.ContentType == ContentTypeDirectory {
+			_, err := zw.Create(ad.Filename)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	// Write agent diagnostics content
+	for _, ad := range agentDiag {
+		if ad.ContentType != ContentTypeDirectory {
+			zf, err := zw.CreateHeader(&zip.FileHeader{
+				Name:     ad.Filename,
+				Method:   zip.Deflate,
+				Modified: ad.Generated,
+			})
+			if err != nil {
+				return err
+			}
+			err = writeRedacted(errOut, zf, ad.Filename, ad)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// Handle unit diagnostics
+	// structure each unit into its own component directory
+	compDirs := make(map[string][]client.DiagnosticUnitResult)
+	for _, ud := range unitDiags {
+		compDir := strings.ReplaceAll(ud.ComponentID, "/", "-")
+		compDirs[compDir] = append(compDirs[compDir], ud)
+	}
+	// write each units diagnostics into its own directory
+	// layout becomes components/<component-id>/<unit-id>/<filename>
+	_, err := zw.Create("components/")
+	if err != nil {
+		return err
+	}
+	for dirName, units := range compDirs {
+		_, err = zw.Create(fmt.Sprintf("components/%s/", dirName))
+		if err != nil {
+			return err
+		}
+		for _, ud := range units {
+			unitDir := strings.ReplaceAll(strings.TrimPrefix(ud.UnitID, ud.ComponentID+"-"), "/", "-")
+			_, err = zw.Create(fmt.Sprintf("components/%s/%s/", dirName, unitDir))
+			if err != nil {
+				return err
+			}
+			if ud.Err != nil {
+				w, err := zw.Create(fmt.Sprintf("components/%s/%s/error.txt", dirName, unitDir))
+				if err != nil {
+					return err
+				}
+				_, err = w.Write([]byte(fmt.Sprintf("%s\n", ud.Err)))
+				if err != nil {
+					return err
+				}
+				continue
+			}
+			for _, fr := range ud.Results {
+				filePath := fmt.Sprintf("components/%s/%s/%s", dirName, unitDir, fr.Name)
+				w, err := zw.CreateHeader(&zip.FileHeader{
+					Name:     filePath,
+					Method:   zip.Deflate,
+					Modified: fr.Generated,
+				})
+				if err != nil {
+					return err
+				}
+				err = writeRedacted(errOut, w, filePath, fr)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func writeRedacted(errOut, w io.Writer, fullFilePath string, fr client.DiagnosticFileResult) error {
```

Comment:
@joshdover, I've tried to incorporate the redaction work you've done as now an archive can be assembled either through an action or from the command line invocation.

Reply:
Great catch 👍